### PR TITLE
RFC document to generate certificates 

### PR DESCRIPTION
**Pull request by prafull01**

### CODE REVIEW

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+Status: Draft
+ 
+Start Date: 19-04-2021
+ 
+Authors: @prafull01, @abhisek-dwivedi, @madhurnawandar
+ 
+# Summary
+ 
+This RFC proposes a method of deploying the CockroachDB Helm chart in a secure mode by generating certificates without using Kubernetes CA.  
+The new method allows users to specify their own CA or use self-generated CA.
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: or if not provided will generate a CA

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+This RFC proposes a method of deploying the CockroachDB Helm chart in a secure mode by generating certificates without using Kubernetes CA.  
+The new method allows users to specify their own CA or use self-generated CA.  
+It handles the creation of Node and Client certificates using the CA.  
+It allows rotation of CA, Node, and Client certificates.  
+It allows specifying a duration for the generated certificates
+ 
+# Motivation
+ 
+Currently, CockroachDB supports 3 ways of cert management
+* Built-in CSR's in Kubernetes: Depends on the user to approve CSR for all the generated certificates manually. CSR is no longer support with Certificates.k8s.io/v1 API and will be deprecated in Kubernetes v1.22. Many Kubernetes distributions like VMware Tanzu, EKS, etc are not allowing Kubernetes CA to sign the CSR's using Kubernetes CA.
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: EKS, etc have deprecated use of Kubernetes CA

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+* Cert-manager: This is the most efficient way of managing certificates, but not everyone uses cert-manager to manage the certificates. Also using cert-manager may be overkill for dev/test environments
+* Manual: The user has to generate the certificates and provide them in the form of secrets. This method puts the overhead of certificate management leading to multiple manual steps for the user
+ 
+CockroachDB user needs a default mechanism of cert management which should work on all the k8s distributions without the need for manual intervention. While cert-manager fits into the requirement, it makes it CockroachDB deployment dependent on external application. This new method of cert-management satisfies the user requirement without the need for 3rd party software like cert-manager.
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: Remove it. Makes CockroachDB deployment dependent...

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+7. Add option specifying CockroachDB to manage rotation of the generated certificates, `tls.certs.generate.rotateCerts` as true/false.  
+  Enabling this option will result in auto-rotation of the certificates, before the expiry.
+ 
+## Helm Input Validation
+ 
+1. If `tls.certs.generate.caProvided` is set to true, then value for `tls.certs.generate.caSecret` must be provided.
+ 
+2. If value for `tls.certs.generate.caSecret` is provided, secret should exist in the CockroachDB install namespace.
+ 
+3. Value for `tls.certs.generate.caCertDuration` should be at least two months greater than the value for `tls.certs.generate.clientCertDuration`
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: Have we finalised 2 month or we need to make cacert duration double of cert duration

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+#### Generate-CA
+* A self-signed CA will be generated
+* Expiry of the certificate will be driven by the Helm value for `tls.certs.generate.caCertDuration`, passed as env variable
+* Contents of the CA certificate will be stored in the default CA secret `cockroachdb-ca`.
+* An annotation `managed-by: cockroachdb` will be added on secret
+* Follow [CA Annotation Workflow](#annotate-ca)
+ 
+#### Generate-Node-Cert
+* A Node certificate will be generated signed by the generated CA or user custom CA.
+* Contents of the Node certificate will be stored in the Node secret `cockroachdb-node`.
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: Need to specify duration of node certificate

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+* An annotation `managed-by: cockroachdb` will be added on secret
+* Node secret will be patched with annotation `resourceVersion` with the value of its current resourceVersion
+* Node secret will be patched with annotation `creationTime` and `duration` with current UTC time and value from `tls.certs.generate.nodeCertDuration`
+ 
+#### Generate-Client-Cert
+* A Client certificate will be generated by signing it with the generated CA
+* Contents of the Client certificate will be stored in the Client secret `cockroachdb-root`
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: Same here

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+  * Do nothing as this `pre-install` job  won't be triggered.
+ 
+### Periodic Rotation scenarios:
+* CA certificate is near expiry: This will be identified using the generation time put on the CA secret. This will lead to a certificate rotation scenario.
+* Node or Client certificate near expiry: This will be identified using the generation time put on the respective secret. This will lead to regeneration of Node or Client certificate
+ 
+### Certificate Rotation scenario:
+ 
+* Only renew CA certificate by generating new CA key and new combined CA along with the old CA.
+* Update CA secret and add CA certificate info in Node and Client secret.
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: What info?

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+`TODO: Need to identify how to generate the SIGHUP signal in all the Nodes for certificate renewal`
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: This todo is at 3 places in this doc

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+* Add annotation on CA secret with the date of rotation.
+* Add annotations on Node and Client certs `to-be-rotated: true`.
+* Do not process Node and Client certificate.
+* On the next scheduled iteration, if `to-be-rotated: true` annotation found, then renew Node or Client certificate.
+* Remove `to-be-rotated: true` from Node secret and Client secret.
+`TODO: Need to identify how to generate the SIGHUP signal in all the Nodes for certificate renewal`
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: 4th instance

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+* User certificate management method is changed from certificate generation to `cert-manager` or `default manual k8s CSR approval`:
+  * Do nothing as this `pre-install` job  won't be triggered.
+ 
+### Periodic Rotation scenarios:
+* CA certificate is near expiry: This will be identified using the generation time put on the CA secret. This will lead to a certificate rotation scenario.
+* Node or Client certificate near expiry: This will be identified using the generation time put on the respective secret. This will lead to regeneration of Node or Client certificate
+ 
+### Certificate Rotation scenario:
+ 
+* Only renew CA certificate by generating new CA key and new combined CA along with the old CA.
```
Comment (madhurtrilio) [inline docs/RFCs/19042021_self_signed_certificates.md]: This info is repeated and already covered in sections above. See if we can remove from somewhere

File: docs/RFCs/19042021_self_signed_certificates.md (hunk 8 above)
Comment (abhishekdwivedi3060) [reply docs/RFCs/19042021_self_signed_certificates.md]: This is added wherever update is required.

File: docs/RFCs/19042021_self_signed_certificates.md (hunk 7 above)
Comment (abhishekdwivedi3060) [reply docs/RFCs/19042021_self_signed_certificates.md]: As per their doc, they update the new CA in the nodes and refresh it and later they sign the new certs using this CA and then again refresh. `CA certificate info` means CA certificate only. Will remove info

File: docs/RFCs/19042021_self_signed_certificates.md (hunk 10 above)
Comment (abhishekdwivedi3060) [reply docs/RFCs/19042021_self_signed_certificates.md]: I think it makes sense to keep one complete flow for CA rotation. Above heading `Rotate-CA` only talks about till CA renewal.

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 11]
```diff
@@ -0,0 +1,240 @@
+Feature Name: Add a method to generate certificates without using Kubernetes CA
```
Comment (madhurnawandar) [inline docs/RFCs/19042021_self_signed_certificates.md]: Change this to `Cert Management in CRDB Helm Chart `

File: docs/RFCs/19042021_self_signed_certificates.md [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+It handles the creation of Node and Client certificates using the CA.  
+It allows rotation of CA, Node, and Client certificates.  
+It allows specifying a duration for the generated certificates
+ 
+# Motivation
+ 
+Currently, CockroachDB supports 3 ways of cert management
+* Built-in CSR's in Kubernetes: Depends on the user to approve CSR for all the generated certificates manually. CSR is no longer support with Certificates.k8s.io/v1 API and will be deprecated in Kubernetes v1.22. Many Kubernetes distributions like VMware Tanzu, EKS, etc are not allowing Kubernetes CA to sign the CSR's using Kubernetes CA.
+* Cert-manager: This is the most efficient way of managing certificates, but not everyone uses cert-manager to manage the certificates. Also using cert-manager may be overkill for dev/test environments
+* Manual: The user has to generate the certificates and provide them in the form of secrets. This method puts the overhead of certificate management leading to multiple manual steps for the user
```
Comment (madhurnawandar) [inline docs/RFCs/19042021_self_signed_certificates.md]: certificate management, leading

add comma

File: docs/RFCs/19042021_self_signed_certificates.md (hunk 1 above)
Comment (abhishekdwivedi3060) [reply docs/RFCs/19042021_self_signed_certificates.md]: Done

File: docs/RFCs/19042021_self_signed_certificates.md (hunk 2 above)
Comment (abhishekdwivedi3060) [reply docs/RFCs/19042021_self_signed_certificates.md]: done

File: docs/RFCs/19042021_self_signed_certificates.md (hunk 3 above)
Comment (abhishekdwivedi3060) [reply docs/RFCs/19042021_self_signed_certificates.md]: done

File: docs/RFCs/19042021_self_signed_certificates.md (hunk 4 above)
Comment (abhishekdwivedi3060) [reply docs/RFCs/19042021_self_signed_certificates.md]: done

File: docs/RFCs/19042021_self_signed_certificates.md (hunk 5 above)
Comment (abhishekdwivedi3060) [reply docs/RFCs/19042021_self_signed_certificates.md]: done

File: docs/RFCs/19042021_self_signed_certificates.md (hunk 6 above)
Comment (abhishekdwivedi3060) [reply docs/RFCs/19042021_self_signed_certificates.md]: done

**Pull request closed without merging**